### PR TITLE
ABC-96 Add paging and server-side search to custom emoji management

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
Add paging and server-side search to custom emoji management.

@jasonblais the styling and behaviour of the next/previous buttons probably needs to be looked at. Please also try the searching and make sure it behaves acceptably.

Can't be tested until mattermost/mattermost-server#8125 goes in.

#### Ticket Link
https://mattermost.atlassian.net/browse/ABC-96

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [x] Has server changes (mattermost/mattermost-server#8125)
- [x] Has redux changes (mattermost/mattermost-redux#369, mattermost/mattermost-redux#367 and mattermost/mattermost-redux#370)
- [x] Has UI changes

### CODE REVIEW

**Comment by jasonblais** [comment]
@jwilander Is this ready for testing?

**Comment by jwilander** [comment]
It will be in a few minutes, just getting it pointed at the right redux commits

**Comment by jwilander** [comment]
I'll also use a script to load a ton of custom emojis onto the spinmint for you

**Comment by jasonblais** [comment]
Sweet, thank you

**Comment by jwilander** [comment]
@jasonblais this one should be ready now too. Same notes as on the other PR:
* Loaded a bunch of custom emoji for you
* Disabled loading all emojis on page load so some posts will display text instead of the emojis, but this won't happen on master and is just so you can test the PR

**Comment by jasonblais** [comment]
@asaadmahmood Can you help review the "Next/Previous" buttons? We can do the same as you did in https://github.com/mattermost/mattermost-webapp/pull/364, with arrows for the Next/Previous buttons, and a little whitespace above them.

image

@asaadmahmood Also, should we display `1 - 100 emoji of 2,903 total` in the UI, or not necessary? I'm 0/5, just seems we do so in most other user/channel lists.

@jwilander Few questions/notes:
 1. Why 100 custom emoji per page? And not, say, 50 or 200?
 2. User not brought to first page after updating search results:
  - On the spinmint, go to Custom Emoji page
  - Go to the bottom of the list, select "Next"
  - Search for an emoji, "zz" --> observed: user still on page 2
  - Remove the search results --> observed: user still on page 2
 3. Search results only show the first 100 emoji
  - On the spinmint, go to Custom Emoji page
  - Search for an emoji, "a" --> observed: only 100 custom emoji shown. No indication that there are more emoji containing the letter "a"
 4. Search by the creator no longer works (regression from `master`)
 5. As a System Admin, unable to delete a custom emoji created by myself (see JS error attached). Doesn't repro with a non-admin account

image

**Comment by jasonblais** [comment]
^I couldn't repro 5 anymore when I tried on a team admin account, and again later on a system admin account, so I'm not sure what happened there.

**Comment by jwilander** [comment]
1. It was chosen arbitrarily, it can be whatever we like
2. I'll fix that
3. Yeah, search for any of our endpoints (users, channels, posts, etc.) only return a limited number of results. We can add paging to it, but we don't have it anywhere else we have searching. It'd be at least 4 mana to do
4. I didn't know we supported searching by creator. That's going to require updates to the API endpoint, probably about 4 mana
5. Not sure what happened, but I'll try and repro anyways

**Comment by jwilander** [comment]
Side note: adding the `1 - 100 emoji of 2,903 total` will require a new API endpoint as well

**Comment by jasonblais** [comment]
@jwilander I think there's a build failure from AM's changes that need to be resolved. I'll set up a new spinmint afterward.

**Comment by jwilander** [comment]
@jasonblais updates made and loaded a bunch of emoji

**Comment by jasonblais** [comment]
Confirmed the following:

0. UI fixed. Ticket to add `1 - 100 emoji of 2,903 total` is here: https://mattermost.atlassian.net/browse/PLT-8665
1. Confirmed, updated to 50
2. Confirmed, fixed
3. Let's leave it for now, given we don't have paging for other search results. We can add if someone asks for it.
4. Posted to UX channel. Will create a separate ticket if it's an important feature.
5. Doesn't repro anymore

**Comment by jwilander** [comment]
@hmhealey @saturninoabril ready for your review

File: components/emoji/emoji_list_item/emoji_list_item.jsx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+        return (
+            <tr className='backstage-list__item'>
+                <td className='emoji-list__name'>
+                    {':' + emoji.name + ':'}
+                </td>
+                <td className='emoji-list__image'>
+                    <span
+                        className='emoticon'
+                        style={{backgroundImage: 'url(' + Client4.getCustomEmojiImageUrl(emoji.id) + ')'}}
```
Comment (saturninoabril) [inline components/emoji/emoji_list_item/emoji_list_item.jsx]: `Client4.getCustomEmojiImageUrl(emoji.id)`
Might qualify to have this in props like `props.emojiImageUrl`?

**Pull request merged**